syntax = "proto3";
import "google/protobuf/empty.proto";
import "google/protobuf/struct.proto";

package kato_proc;

service KatoEngine {
  rpc GetName(google.protobuf.Empty) returns (StandardResponse) {}
  rpc Observe(Observation) returns (ObservationResponse) {}
  rpc Learn(google.protobuf.Empty) returns (StandardResponse) {}
  rpc StartSleeping(google.protobuf.Empty) returns (StandardResponse) {}
  rpc StopSleeping(google.protobuf.Empty) returns (StandardResponse) {}
  rpc StartPredicting(google.protobuf.Empty) returns (StandardResponse) {}
  rpc StopPredicting(google.protobuf.Empty) returns (StandardResponse) {}
  rpc UpdateGenes(GeneData) returns (StandardResponse) {}
  rpc IncrementRecallThreshold(Increment) returns (StandardResponse) {}
  rpc GetGene(Gene) returns (StandardResponse) {}
  rpc ShowStatus(google.protobuf.Empty) returns (StandardResponse) {}
  rpc GetTime(google.protobuf.Empty) returns (StandardResponse) {}
  rpc GetPerceptData(google.protobuf.Empty) returns (StandardResponse) {}
  rpc GetCognitionData(google.protobuf.Empty) returns (CognitionResponse) {}
  rpc GetModel(Name) returns (StandardResponse) {}
  rpc DeleteModel(Name) returns (StandardResponse) {}
  rpc UpdateModel(ModelUpdate) returns (StandardResponse) {}
  rpc GetVector(Name) returns (StandardResponse) {}
  rpc GetWorkingMemory(google.protobuf.Empty) returns (StandardResponse) {}
  rpc GetAllPredictions(google.protobuf.Empty) returns (StandardResponse) {}
  rpc GetPredictions(UniqueId) returns (StandardResponse) {}
  rpc ClearWorkingMemory(google.protobuf.Empty) returns (StandardResponse) {}
  rpc ClearAllMemory(google.protobuf.Empty) returns (StandardResponse) {}
  rpc Ping(google.protobuf.Empty) returns (StandardResponse) {}
  rpc TargetClass(TargetClassRequest) returns (StandardResponse) {}
  rpc TargetClear(google.protobuf.Empty) returns (StandardResponse) {}
}

message StandardResponse {
  enum Status {
    OKAY = 0;
    FAILED = 1;
  }
  Status status = 1;
  string id = 2;
  int32 interval = 3;
  float time_stamp = 4;
  oneof response_oneof {
    string message = 5;
    google.protobuf.Struct response = 6;
  }
}

message WorkingMemoryResponse {
  enum Status {
    OKAY = 0;
    FAILED = 1;
  }
  Status status = 1;
  string id = 2;
  int32 interval = 3;
  float time_stamp = 4;
  oneof response_oneof {
    string message = 5;
    google.protobuf.ListValue response = 6;
  }
}


message ObservationResponse {
  enum Status {
    OKAY = 0;
    FAILED = 1;
  }
  Status status = 1;
  string id = 2;
  int32 interval = 3;
  float time_stamp = 4;
  string message = 5;
  string unique_id = 6;
}

message ModelObjectResponse {
   enum Status {
    OKAY = 0;
    FAILED = 1;
  }
  Status status = 1;
  string id = 2;
  int32 interval = 3;
  float time_stamp = 4;
  oneof response_oneof {
    string message = 5;
    ModelObject response = 6;
  }
}

message Observation {
  repeated string strings = 1;
  repeated google.protobuf.ListValue vectors = 2;
  map <string, float> emotives = 3;
  map <string, string> metadata = 4;
  string command = 5;
  string unique_id = 6;
  repeated string path = 7;
  repeated string symbols = 8;
  repeated Prediction predictions = 9;
  string source = 10;
}

message Prediction {
    string name = 1;
    string type = 2;
    int32 frequency = 3;
    float potential = 4;
    float confidence = 5;
    float evidence = 6;
    float fragmentation = 7;
    float hamiltonian = 8;
    float grand_hamiltonian = 9;
    float entropy = 10;
    float itfdf_similarity = 11;
    float confluence = 12;
    float snr = 13;
    float similarity = 14;
    repeated string missing = 16;
    repeated string matches = 17;
    repeated string extras = 18;
    repeated google.protobuf.ListValue past = 19;
    repeated google.protobuf.ListValue present = 20;
    repeated google.protobuf.ListValue future = 21;
    map <string, float> emotives = 22;
}

message ModelObject {
    string name = 1;
    int32 frequency = 2;
    int32 length = 3;
    repeated google.protobuf.Struct emotives = 4;
    repeated google.protobuf.ListValue sequence = 5;
}

message ModelUpdate {
  string name = 1;
  int32 frequency = 2;
  repeated google.protobuf.Struct emotives = 3;
}

message GeneData {
  map <string, float> genes = 1;
}

message Name {
  string name = 1;
}

message Gene {
  string gene = 1;
}

message Increment {
  int32 value = 1;
}

message CognitionObject {
  repeated Prediction predictions = 1;
  map <string, float> emotives = 2;
  repeated string symbols = 3;
  string command = 4;
  map <string, string> metadata = 5;
  repeated string path = 6;
  repeated string strings = 7;
  repeated google.protobuf.ListValue vectors = 8;
  repeated google.protobuf.ListValue working_memory = 9;
}

message CognitionResponse {
  enum Status {
    OKAY = 0;
    FAILED = 1;
  }
  Status status = 1;
  string id = 2;
  int32 interval = 3;
  float time_stamp = 4;
  oneof response_oneof {
    string message = 5;
    CognitionObject response = 6;
  }
}

message UniqueId {
  string unique_id = 1;
}

message TargetClassRequest {
  string target_class = 1;
}

